### PR TITLE
Add multi-connection support

### PR DESCRIPTION
**Pull request by GTwhy**

### CODE REVIEW

File: src/memory_region/mod.rs [hunk 1]
```diff
@@ -20,8 +23,14 @@ pub trait MrAccess: Sync + Send + Debug {
     /// Get the remote key
     fn rkey(&self) -> u32;
 
-    /// Get the access of mr
-    fn access(&self) -> ibv_access_flags;
+    /// Get the `ibv_access_flags` of mr
+    fn ibv_access(&self) -> ibv_access_flags;
```
Comment (rogercloud) [inline src/memory_region/mod.rs:27]: Should we return a rdma_sys type?

File: src/mr_allocator.rs [hunk 2]
```diff
@@ -72,12 +72,24 @@ static mut RDMA_EXTENT_HOOKS: extent_hooks_t = extent_hooks_t {
...
 };
 
+/// Combination of `ProtectionDomain`'s ptr and `ibv_access_falgs` as a key.
+#[derive(PartialEq, Eq, Hash)]
+pub(crate) struct AccessPDKey(u128);
+
+impl AccessPDKey {
+    /// Cast the `ProtectionDomain`'s ptr and `ibv_access_flags` into a `u128`
+    #[allow(clippy::as_conversions)]
+    pub(crate) fn new(pd: &Arc<ProtectionDomain>, access: ibv_access_flags) -> Self {
```
Comment (rogercloud) [inline src/mr_allocator.rs:82]: Why don't we merge protection domain id and flags? And why doesn't a tuple work here?

Comment (rogercloud) [reply src/mr_allocator.rs:82]: A tuple means `(protection domain id, flags)`

File: src/mr_allocator.rs [hunk 3]
```diff
@@ -222,7 +251,7 @@ impl MrAllocator {
 
         match self.strategy {
             MRManageStrategy::Jemalloc => {
-                if access == self.default_access{
+                if access == self.default_access && pd.inner_pd == self.default_pd.inner_pd {
```
Comment (rogercloud) [inline src/mr_allocator.rs:254]: can we also put default access and pd in the map? So the code is cleaner.

File: examples/client.rs [hunk 4]
```diff
@@ -102,15 +102,25 @@ async fn request_then_write_with_imm(rdma: &Rdma) -> io::Result<()> {
...
+    let rdma = RdmaBuilder::default().connect(addr).await.unwrap();
     println!("connected");
     send_data_to_server(&rdma).await.unwrap();
     send_data_with_imm_to_server(&rdma).await.unwrap();
     send_lmr_to_server(&rdma).await.unwrap();
     request_then_write(&rdma).await.unwrap();
     request_then_write_with_imm(&rdma).await.unwrap();
     println!("client done");
+
+    // create new `Rdma`s (connections) that has the same `mr_allocator` and `event_listener` as parent
```
[the remark this answers is not in the document]
Comment (GTwhy) [reply examples/client.rs:115]: We can, but maybe not necessary for now.

We support different pd and access and I don't know what kind of scenario needs a new allocator, it's not good for efficiency or safety.

New event_listener maybe useful if it supports more than one poll strategies, so maybe we can add this latter.

Of course, we can also add support in this version. What do you think?

File: src/memory_region/mod.rs (hunk 1 above)
Comment (GTwhy) [reply src/memory_region/mod.rs:27]: Exposing rdma_sys type is usually not a good idea, but we can use it to do some access checks quickly without the need of importing rdma_sys type.

High-level access API is also exposed, and it has a little overhead for converting. Users can choose one of them.

File: src/mr_allocator.rs (hunk 2 above)
Comment (GTwhy) [reply src/mr_allocator.rs:82]: Oh, tuple is certainly workable, why I wrote this ┑(￣。。￣)┍

File: src/mr_allocator.rs (hunk 3 above)
Comment (rogercloud) [reply src/mr_allocator.rs:254]: At least we can share some code to deduplicate.

File: examples/client.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (rogercloud) [reply examples/client.rs:115]: We can add the feature later, don't forget to file an issue to track.

File: src/memory_region/mod.rs (hunk 1 above)
Comment (rogercloud) [reply src/memory_region/mod.rs:27]: Can we find a way to speed up the converting between these two types?

Comment (rogercloud) [reply src/memory_region/mod.rs:27]: One way is to enumerate all the possible combinations "automatically" via macro. The drawback of this method is the code size.

File: examples/client.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (GTwhy) [reply examples/client.rs:115]: Tracked by #84

**Review by rogercloud — APPROVED** [review]

**Pull request merged**